lib/pure/ioselectors.nim

```diff
@@ -0,0 +1,1849 @@
+  # 2) You can't have one event, which you can use with many kqueue handles.
+  # So decision was made in favor of the pipes
+  #
+  when bsdPlatform:
+    const
+      MAX_KQUEUE_CHANGE_EVENTS = 64
+      MAX_KQUEUE_RESULT_EVENTS = 64
+
+    when hasThreadSupport:
+      type
+        SelectorImpl[T] = object
+          kqFD : cint
+          maxFD : uint
+          changesTable: array[MAX_KQUEUE_CHANGE_EVENTS, KEvent]
+          changesCount: int
+          fds: ptr SharedArray[SelectorKey[T]]
+          count: int
+          changesLock: Lock
+    else:
+      type
+        SelectorImpl[T] = object
+          kqFD : cint
+          maxFD : uint
+          changesTable: array[MAX_KQUEUE_CHANGE_EVENTS, KEvent]
+          changesCount: int
+          fds: seq[SelectorKey[T]]
+          count: int
+
+    when hasThreadSupport:
+      type Selector*[T] = ptr SelectorImpl[T]
+    else:
+      type Selector*[T] = ref SelectorImpl[T]
+
+    type
+      SelectEventImpl = object
+        rfd: cint
+        wfd: cint
+    # SelectEvent is declared as `ptr` to be placed in `shared memory`,
+    # so you can share one SelectEvent handle between threads.
+    type SelectEvent* = ptr SelectEventImpl
+
+    proc newSelector*[T](): Selector[T] =
+      var maxFD = getMaxFds()
+      var kqFD = kqueue()
+      if kqFD < 0:
+        raiseOsError(osLastError())
+      when hasThreadSupport:
+        result = cast[Selector[T]](allocShared0(sizeof(SelectorImpl[T])))
+        result.kqFD = kqFD
+        result.maxFD = maxFD.uint
+        result.fds = allocSharedArray[SelectorKey[T]](maxFD)
+        initLock(result.changesLock)
+      else:
+        result = Selector[T](kqFD: kqFD, maxFD: maxFD.uint)
+        result.fds = newSeq[SelectorKey[T]](maxFD)
+
+    proc close*[T](s: Selector[T]) =
+      if posix.close(s.kqFD) != 0:
+        raiseOSError(osLastError())
+      when hasThreadSupport:
+        deinitLock(s.changesLock)
+        deallocSharedArray(s.fds)
+        deallocShared(cast[pointer](s))
+
+    when hasThreadSupport:
+      template withChangeLock[T](s: Selector[T], body: untyped) =
+        acquire(s.changesLock)
+        {.locks: [s.changesLock].}:
+          try:
+            body
+          finally:
+            release(s.changesLock)
+    else:
+      template withChangeLock(s, body: untyped) =
+        body
+
+    template modifyKQueue[T](s: Selector[T], nident: uint, nfilter: cshort,
+                             nflags: cushort, nfflags: cuint, ndata: int,
+                             nudata: pointer) =
+      mixin withChangeLock
+      s.withChangeLock():
+        s.changesTable[s.changesCount] = KEvent(ident: nident,
+                                                filter: nfilter, flags: nflags,
+                                                fflags: nfflags, data: ndata,
+                                                udata: nudata)
+        inc(s.changesCount)
+        if s.changesCount == MAX_KQUEUE_CHANGE_EVENTS:
```

Comment:
Better move this check before you attempt to add to 's.changesTable', if possible. Allows 1 more entry, avoids RAW pipeline stalls. (Are these still a thing on modern CPUs? Should be taken care of by store forwarding, right?)